File: packages/ts/src/vrf.ts
/**
 * VRF precompile (`0x1101`) helpers.
 *
 * This precompile is selectorless: calldata is a 32-byte big-endian block
 * height followed by a caller-chosen domain tag. Successful return data is
 * exactly 32 bytes.
 */

import { PRECOMPILE_ADDRESSES } from "./consts.js";
import { bytesToHex, concatBytes, hexToBytes } from "./crypto/bytes.js";

export class VrfCallError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "VrfCallError";
  }
}

export const VRF_OUTPUT_BYTES = 32 as const;
export const VRF_DOMAIN_TAG_MAX_BYTES = 256 as const;
export const VRF_HEIGHT_NOT_FINALIZED_REVERT = "vrf: height not finalized" as const;

export type VrfDomainTagInput = string | Uint8Array | readonly number[];

/** Return the VRF precompile address (`0x1101`) as lower-case hex. */
export function vrfAddressHex(): string {
  return PRECOMPILE_ADDRESSES.VRF.toLowerCase();
}

/**
 * Encode selectorless VRF calldata: `uint256 blockHeight || domainTag`.
 *
 * @param blockHeight finalized block height to read randomness from.
 * @param domainTag independent namespace for the consumer, up to 256 bytes.
 */
export function encodeVrfEvaluateCalldata(
  blockHeight: bigint | number | string,
  domainTag: VrfDomainTagInput = new Uint8Array(),
): string {
  const height = parseUint64(blockHeight, "blockHeight");
  const tag = normalizeDomainTag(domainTag);
  if (tag.length > VRF_DOMAIN_TAG_MAX_BYTES) {
    throw new VrfCallError(`domainTag exceeds ${VRF_DOMAIN_TAG_MAX_BYTES} bytes`);
  }
  return bytesToHex(concatBytes(uint256Word(height), tag));
}

/** Decode a successful VRF return payload into 32 bytes. */
export function decodeVrfOutput(output: string | Uint8Array | readonly number[]): Uint8Array {
  const bytes = toBytes(output, "output");
  if (bytes.length !== VRF_OUTPUT_BYTES) {
    throw new VrfCallError(`VRF output must be ${VRF_OUTPUT_BYTES} bytes, got ${bytes.length}`);
  }
  return bytes;
}

function normalizeDomainTag(value: VrfDomainTagInput): Uint8Array {
  if (typeof value === "string") {
    if (value.startsWith("0x") || value.startsWith("0X")) return hexToBytes(value, "domainTag");
    return new TextEncoder().encode(value);
  }
  return value instanceof Uint8Array ? value : Uint8Array.from(value);
}

function parseUint64(value: bigint | number | string, label: string): bigint {
  let parsed: bigint;
  if (typeof value === "bigint") {
    parsed = value;
  } else if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) throw new VrfCallError(`${label} must be a safe integer`);
    parsed = BigInt(value);
  } else if (value.startsWith("0x") || value.startsWith("0X")) {
    parsed = BigInt(value);
  } else {
    if (!/^[0-9]+$/.test(value)) throw new VrfCallError(`${label} must be a non-negative integer`);
    parsed = BigInt(value);
  }
  if (parsed < 0n || parsed > (1n << 64n) - 1n) {
    throw new VrfCallError(`${label} out of uint64 range`);
  }
  return parsed;
}

function uint256Word(value: bigint): Uint8Array {
  const out = new Uint8Array(32);
  let rest = value;
  for (let i = 31; i >= 0 && rest > 0n; i--) {
    out[i] = Number(rest & 0xffn);
    rest >>= 8n;
  }
  return out;
}

function toBytes(value: string | Uint8Array | readonly number[], label: string): Uint8Array {
  if (typeof value === "string") return hexToBytes(value, label);
  return value instanceof Uint8Array ? value : Uint8Array.from(value);
}

File: crates/core-sdk/src/vrf.rs
//! VRF precompile (`0x1101`) calldata helpers.
//!
//! Mirrors `packages/ts/src/vrf.ts`. The precompile is selectorless:
//! calldata is a 32-byte big-endian block height followed by a
//! caller-chosen domain tag (up to 256 bytes). Successful return data is
//! exactly 32 bytes. Reading an unfinalized height reverts with
//! [`VRF_HEIGHT_NOT_FINALIZED_REVERT`].
//!
//! The historical leader seed is also exposed as
//! `lyth_getRoundCertificate(round).signature` (the ML-DSA-65 leader-seed
//! digest); see [`crate::RoundCertificateResponse`].

use crate::consts::precompile_addresses::VRF;

/// VRF precompile address (`0x1101`).
pub const VRF_ADDRESS: [u8; 20] = VRF;

/// Successful VRF return-data width.
pub const VRF_OUTPUT_BYTES: usize = 32;
/// Maximum caller domain-tag width.
pub const VRF_DOMAIN_TAG_MAX_BYTES: usize = 256;
/// Revert message returned when the requested height has not finalized.
pub const VRF_HEIGHT_NOT_FINALIZED_REVERT: &str = "vrf: height not finalized";

/// Error from the VRF calldata helpers.
#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
pub enum VrfCallError {
    /// The domain tag exceeds [`VRF_DOMAIN_TAG_MAX_BYTES`].
    #[error("domainTag exceeds {VRF_DOMAIN_TAG_MAX_BYTES} bytes")]
    DomainTagTooLong,
    /// VRF return data was not exactly [`VRF_OUTPUT_BYTES`] bytes.
    #[error("VRF output must be {VRF_OUTPUT_BYTES} bytes, got {0}")]
    BadOutputLen(usize),
}

/// Return the VRF precompile address (`0x1101`) as lower-case hex.
#[must_use]
pub fn vrf_address_hex() -> String {
    let mut out = String::with_capacity(42);
    out.push_str("0x");
    for b in &VRF_ADDRESS {
        out.push_str(&format!("{b:02x}"));
    }
    out
}

/// Encode selectorless VRF calldata: `uint256 block_height || domain_tag`.
///
/// `block_height` is the finalized height to read randomness from;
/// `domain_tag` is an independent namespace for the consumer.
///
/// # Errors
/// [`VrfCallError::DomainTagTooLong`] when `domain_tag` exceeds
/// [`VRF_DOMAIN_TAG_MAX_BYTES`].
pub fn encode_vrf_evaluate_calldata(
    block_height: u64,
    domain_tag: &[u8],
) -> Result<Vec<u8>, VrfCallError> {
    if domain_tag.len() > VRF_DOMAIN_TAG_MAX_BYTES {
        return Err(VrfCallError::DomainTagTooLong);
    }
    let mut out = Vec::with_capacity(32 + domain_tag.len());
    let mut word = [0u8; 32];
    word[24..32].copy_from_slice(&block_height.to_be_bytes());
    out.extend_from_slice(&word);
    out.extend_from_slice(domain_tag);
    Ok(out)
}

/// Encode selectorless VRF calldata as lower-case `0x` hex.
///
/// # Errors
/// [`VrfCallError::DomainTagTooLong`] when `domain_tag` is too long.
pub fn encode_vrf_evaluate_calldata_hex(
    block_height: u64,
    domain_tag: &[u8],
) -> Result<String, VrfCallError> {
    Ok(calldata_to_hex(&encode_vrf_evaluate_calldata(
        block_height,
        domain_tag,
    )?))
}

/// Decode a successful VRF return payload into its 32 randomness bytes.
///
/// # Errors
/// [`VrfCallError::BadOutputLen`] when `output` is not 32 bytes.
pub fn decode_vrf_output(output: &[u8]) -> Result<[u8; VRF_OUTPUT_BYTES], VrfCallError> {
    if output.len() != VRF_OUTPUT_BYTES {
        return Err(VrfCallError::BadOutputLen(output.len()));
    }
    let mut out = [0u8; VRF_OUTPUT_BYTES];
    out.copy_from_slice(output);
    Ok(out)
}

/// Format raw calldata as `0x...` hex.
#[must_use]
pub fn calldata_to_hex(calldata: &[u8]) -> String {
    let mut out = String::with_capacity(2 + calldata.len() * 2);
    out.push_str("0x");
    for b in calldata {
        out.push_str(&format!("{b:02x}"));
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn vrf_address_matches_consts() {
        assert_eq!(VRF_ADDRESS, crate::consts::precompile_addresses::VRF);
        assert_eq!(
            vrf_address_hex(),
            "0x0000000000000000000000000000000000001101"
        );
    }

    #[test]
    fn encodes_selectorless_finalized_height_calldata() {
        // Golden vector reused from packages/ts/tests/vrf.test.ts.
        assert_eq!(
            encode_vrf_evaluate_calldata_hex(42, b"dice").unwrap(),
            "0x000000000000000000000000000000000000000000000000000000000000002a64696365",
        );
        assert_eq!(
            encode_vrf_evaluate_calldata_hex(0x2a, &[0x01, 0x02]).unwrap(),
            "0x000000000000000000000000000000000000000000000000000000000000002a0102",
        );
    }

    #[test]
    fn rejects_oversized_domain_tag() {
        let tag = vec![0u8; VRF_DOMAIN_TAG_MAX_BYTES + 1];
        assert_eq!(
            encode_vrf_evaluate_calldata(1, &tag),
            Err(VrfCallError::DomainTagTooLong)
        );
        // Exactly at the cap is accepted.
        assert!(encode_vrf_evaluate_calldata(1, &vec![0u8; VRF_DOMAIN_TAG_MAX_BYTES]).is_ok());
    }

    #[test]
    fn decodes_32_byte_output() {
        let out = decode_vrf_output(&[0xab; 32]).unwrap();
        assert_eq!(out, [0xab; 32]);
        assert_eq!(
            decode_vrf_output(&[0x12, 0x34]),
            Err(VrfCallError::BadOutputLen(2))
        );
    }
}
